### PR TITLE
made 2 states (designer or not) to 3 states (designer, not, unknown)

### PR DESCRIPTION
**Pull request by heejaechang**
when it is unknown status, we will leave designer attribute as it used to be. and not touch anything.

one can go into unknown state if there is no explict designer attribute found but there are some types
with errors.

this fix devdiv 218940 - https://devdiv.visualstudio.com/DevDiv/_workitems?id=218940&_a=edit

### CODE REVIEW

**Comment by heejaechang** [comment]
test vsi please

**Comment by heejaechang** [comment]
I fixed this bug since it is marked as update 3 bug, but not sure whether it is actually update 3 bug or not.

**Comment by heejaechang** [comment]
@mavasani @srivatsn @dotnet/roslyn-analysis can you take a look?

File: src/VisualStudio/Core/Def/Implementation/DesignerAttribute/AbstractDesignerAttributeIncrementalAnalyzer.cs [hunk 1]
```diff
@@ -185,34 +196,30 @@ private static string GetArgumentString(TypedConstant argument)
         }
 
         private async System.Threading.Tasks.Task RegisterDesignerAttributeAndSaveStateAsync(
-            Document document, VersionStamp textVersion, VersionStamp semanticVersion, string designerAttributeArgument,
-            Data existingData, CancellationToken cancellationToken)
+            Document document, VersionStamp textVersion, VersionStamp semanticVersion, Optional<string> designerAttributeArgumentOpt, CancellationToken cancellationToken)
         {
-            if (existingData != null && string.Equals(existingData.DesignerAttributeArgument, designerAttributeArgument, StringComparison.OrdinalIgnoreCase))
```
Comment (heejaechang) [inline src/VisualStudio/Core/Def/Implementation/DesignerAttribute/AbstractDesignerAttributeIncrementalAnalyzer.cs]: this is wrong, but sure why this is here..

**Comment by heejaechang** [comment]
test vsi please

**Comment by heejaechang** [comment]
@dotnet-bot retest windows_vsi_p3_open_prtest please
// Previous failure: http://dotnet-ci.cloudapp.net/job/Private/job/dotnet_roslyn-internal/job/master/job/windows_vsi_p3_open_prtest/22/
// Retest reason: github issue

**Comment by heejaechang** [comment]
ping?

File: src/VisualStudio/Core/Def/Implementation/DesignerAttribute/AbstractDesignerAttributeIncrementalAnalyzer.cs [hunk 2]
```diff
@@ -102,6 +101,8 @@ public bool NeedsReanalysisOnOptionChanged(object sender, OptionChangedEventArgs
             INamedTypeSymbol designerAttribute = null;
             SemanticModel model = null;
 
+            var documentHasError = false;
```
Comment (heejaechang) [inline src/VisualStudio/Core/Def/Implementation/DesignerAttribute/AbstractDesignerAttributeIncrementalAnalyzer.cs]: this is actual fix.

File: src/VisualStudio/Core/Def/Implementation/DesignerAttribute/AbstractDesignerAttributeIncrementalAnalyzer.cs [hunk 3]
```diff
@@ -139,14 +141,20 @@ public bool NeedsReanalysisOnOptionChanged(object sender, OptionChangedEventArgs
                     // walk up type chain
                     foreach (var type in definedType.GetBaseTypesAndThis())
                     {
+                        if (type.IsErrorType())
```
Comment (heejaechang) [inline src/VisualStudio/Core/Def/Implementation/DesignerAttribute/AbstractDesignerAttributeIncrementalAnalyzer.cs]: and here

File: src/VisualStudio/Core/Def/Implementation/DesignerAttribute/AbstractDesignerAttributeIncrementalAnalyzer.cs [hunk 4]
```diff
@@ -160,7 +168,10 @@ public bool NeedsReanalysisOnOptionChanged(object sender, OptionChangedEventArgs
                 }
             }
 
-            await RegisterDesignerAttributeAndSaveStateAsync(document, textVersion, semanticVersion, designerAttributeArgument, existingData, cancellationToken).ConfigureAwait(false);
+            // we checked all types in the document, but couldn't find designer attribute, but we can't say this document doesn't have designer attribute
+            // if the document also contains some errors.
+            var designerAttributeArgumentOpt = documentHasError ? new Optional<string>() : new Optional<string>(designerAttributeArgument);
```
Comment (heejaechang) [inline src/VisualStudio/Core/Def/Implementation/DesignerAttribute/AbstractDesignerAttributeIncrementalAnalyzer.cs]: and here

File: src/VisualStudio/Core/Def/Implementation/DesignerAttribute/AbstractDesignerAttributeIncrementalAnalyzer.cs [hunk 5]
```diff
@@ -185,34 +196,30 @@ private static string GetArgumentString(TypedConstant argument)
         }
 
         private async System.Threading.Tasks.Task RegisterDesignerAttributeAndSaveStateAsync(
-            Document document, VersionStamp textVersion, VersionStamp semanticVersion, string designerAttributeArgument,
-            Data existingData, CancellationToken cancellationToken)
+            Document document, VersionStamp textVersion, VersionStamp semanticVersion, Optional<string> designerAttributeArgumentOpt, CancellationToken cancellationToken)
         {
-            if (existingData != null && string.Equals(existingData.DesignerAttributeArgument, designerAttributeArgument, StringComparison.OrdinalIgnoreCase))
+            if (!designerAttributeArgumentOpt.HasValue)
```
Comment (heejaechang) [inline src/VisualStudio/Core/Def/Implementation/DesignerAttribute/AbstractDesignerAttributeIncrementalAnalyzer.cs]: and here. all others are just cleaning code up.

**Comment by mavasani** [comment]
:+1:

**Comment by MattGertz** [comment]
Approved pending one more code review.

**Comment by srivatsn** [comment]
:+1:

**Pull request merged**